File: Competitive Programming 3/Section 1.4/Real Life Easier/187 - Transaction Processing/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

map<int,string> accounts;
vector<pair<int,int> > transactionEntries; //<account,number>

void process(int sequenceNumber,int acum){
    cout << "*** Transaction " << sequenceNumber << " is out of balance ***\n";
    for(auto entry : transactionEntries){
        string name = accounts[entry.first];
        cout << entry.first << " " << name << string(30-name.size()+1,' ') << setw(10) << fixed << setprecision(2) << (float)entry.second/100 << "\n";
    } 
    cout << "999 Out of Balance" << string(17,' ') << setw(10) << fixed << setprecision(2) << (float)-acum/100 << "\n\n";
}

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    string line;
    while(getline(cin,line)&&line.substr(0,3)!="000"){
        int account = stoi(line.substr(0,3));
        accounts[account]=line.substr(3,line.size()-3);
    }
    int lastS=0,acum=0;
    while(getline(cin,line)&&line.substr(0,3)!="000"){
        int sequenceNumber = stoi(line.substr(0,3));
        if(lastS!=sequenceNumber){
            if(acum){
                process(lastS,acum);
                acum=0;
            }
            lastS = sequenceNumber;
            transactionEntries.clear();
        }
        int account = stoi(line.substr(3,3));
        int number = stoi(line.substr(6,line.size()-6));
        acum+=number;
        transactionEntries.push_back(make_pair(account,number));
    }
    if(acum)
        process(lastS,acum);
    return 0;
}

File: Competitive Programming 3/Section 1.4/Real Life Easier/187 - Transaction Processing/Java/Main.java
import java.util.*;

public class Main{
    static HashMap<Integer,String> accounts = new HashMap<Integer,String>();
    static class Entry{
        public int account,number;
        public Entry(int account,int number){
            this.account=account;
            this.number=number;
        }
    }
    static Vector<Entry> transactionEntries = new Vector<Entry>();
    static void process(int sequenceNumber,int acum){
        System.out.printf("*** Transaction %d is out of balance ***\n",sequenceNumber);
        for(Entry entry: transactionEntries){
            String name = accounts.get(entry.account);
            String amountS = String.format("%.2f",(float)entry.number/100);
            System.out.printf("%d %s%s%s\n",entry.account,name,new String(new char[30-name.length()+11-amountS.length()]).replace('\0',' '),amountS);
        }
        String amountS = String.format("%.2f",(float)-acum/100);
        System.out.printf("999 Out of Balance%s%s\n\n",new String(new char[27-amountS.length()]).replace('\0',' '),amountS);
    }
    static Scanner sc = new Scanner(System.in);
    public static void main(String[]args){
        String line;
        while(!(line = sc.nextLine()).substring(0,3).equals("000")){
            int account = Integer.parseInt(line.substring(0,3));
            accounts.put(account,line.substring(3,line.length()));
        }
        int lastS=0,acum=0;
        while(!(line = sc.nextLine()).substring(0,3).equals("000")){
            int sequenceNumber = Integer.parseInt(line.substring(0,3));
            if(lastS!=sequenceNumber){
                if(acum!=0){
                    process(lastS,acum);
                    acum=0;
                }
                lastS = sequenceNumber;
                transactionEntries.clear();
            }
            int account = Integer.parseInt(line.substring(3,6));
            int number = Integer.parseInt(line.substring(6,line.length()).trim());
            acum += number;
            transactionEntries.add(new Entry(account,number));
        }
        if(acum!=0)
            process(lastS,acum);
    }
}